File: src/game/achievement-service.ts
import type { PlayerSession } from './player.js';
import { ACHIEVEMENTS, grantAchievement } from './achievements.js';
import type { SendFn } from './commands.js';

export function checkAchievements(
  player: PlayerSession,
  send: SendFn,
  triggers: {
    kill?: string;
    room?: string;
    gold?: number;
    level?: number;
    duelWin?: boolean;
  },
): void {
  const earned = player.data.achievements;

  if (triggers.kill && player.data.kills === 1) {
    notify(player, send, grantAchievement(earned, 'first_blood'));
  }

  if (triggers.kill?.includes('goblin') && (player.data.goblinKills ?? 0) >= 10) {
    notify(player, send, grantAchievement(earned, 'goblin_hunter'));
  }

  if (triggers.kill === 'goblin_chief') {
    notify(player, send, grantAchievement(earned, 'chief_slayer'));
  }

  if (triggers.room === 'crystal_cavern') {
    notify(player, send, grantAchievement(earned, 'explorer'));
  }

  if (triggers.gold !== undefined && triggers.gold >= 500) {
    notify(player, send, grantAchievement(earned, 'wealthy'));
  }

  if (triggers.level !== undefined && triggers.level >= 10) {
    notify(player, send, grantAchievement(earned, 'veteran'));
  }

  if (triggers.duelWin) {
    notify(player, send, grantAchievement(earned, 'duelist'));
  }
}

function notify(
  player: PlayerSession,
  send: SendFn,
  achievement: ReturnType<typeof grantAchievement>,
): void {
  if (!achievement) return;
  send(player, {
    type: 'output',
    text: `*** ACHIEVEMENT: ${achievement.name} ***\n${achievement.description}`,
    style: 'quest',
  });
  if (achievement.title) {
    player.data.title = achievement.title;
    send(player, { type: 'output', text: `Title earned: ${achievement.title}`, style: 'loot' });
  }
  send(player, { type: 'bell' });
}

export function formatAchievements(player: PlayerSession): string {
  if (player.data.achievements.length === 0) {
    return 'No achievements yet. Explore, fight, and conquer!';
  }
  const lines = ['-- Achievements --'];
  for (const id of player.data.achievements) {
    const a = ACHIEVEMENTS[id];
    if (a) lines.push(`  ✓ ${a.name} — ${a.description}`);
  }
  if (player.data.title) lines.push(`\nTitle: ${player.data.title}`);
  return lines.join('\n');
}

File: crates/realm-core/src/achievement_service.rs
use realm_protocol::{OutputStyle, ServerMessage};

use crate::achievements::{grant_achievement, AchievementDef};
use crate::player::PlayerSession;

#[derive(Debug, Clone, Default)]
pub struct AchievementTriggers {
    pub kill: Option<String>,
    pub room: Option<String>,
    pub gold: Option<i32>,
    pub level: Option<u32>,
    pub duel_win: Option<bool>,
}

pub fn check_achievements<F>(player: &mut PlayerSession, mut send: F, triggers: AchievementTriggers)
where
    F: FnMut(&str, ServerMessage),
{
    if triggers.kill.is_some() && player.data.kills == 1 {
        let achievement = grant_achievement(&mut player.data.achievements, "first_blood");
        notify(player, &mut send, achievement);
    }

    if triggers
        .kill
        .as_ref()
        .is_some_and(|k| k.contains("goblin"))
        && player.data.goblin_kills >= 10
    {
        let achievement = grant_achievement(&mut player.data.achievements, "goblin_hunter");
        notify(player, &mut send, achievement);
    }

    if triggers.kill.as_deref() == Some("goblin_chief") {
        let achievement = grant_achievement(&mut player.data.achievements, "chief_slayer");
        notify(player, &mut send, achievement);
    }

    if triggers.room.as_deref() == Some("crystal_cavern") {
        let achievement = grant_achievement(&mut player.data.achievements, "explorer");
        notify(player, &mut send, achievement);
    }

    if triggers.gold.is_some_and(|g| g >= 500) {
        let achievement = grant_achievement(&mut player.data.achievements, "wealthy");
        notify(player, &mut send, achievement);
    }

    if triggers.level.is_some_and(|l| l >= 10) {
        let achievement = grant_achievement(&mut player.data.achievements, "veteran");
        notify(player, &mut send, achievement);
    }

    if triggers.duel_win == Some(true) {
        let achievement = grant_achievement(&mut player.data.achievements, "duelist");
        notify(player, &mut send, achievement);
    }
}

fn notify<F>(player: &mut PlayerSession, send: &mut F, achievement: Option<AchievementDef>)
where
    F: FnMut(&str, ServerMessage),
{
    let Some(achievement) = achievement else {
        return;
    };

    let username = player.username().to_string();

    send(
        &username,
        ServerMessage::Output {
            text: format!(
                "*** ACHIEVEMENT: {} ***\n{}",
                achievement.name, achievement.description
            ),
            style: Some(OutputStyle::Quest),
        },
    );

    if let Some(title) = achievement.title {
        player.data.title = Some(title.to_string());
        send(
            &username,
            ServerMessage::Output {
                text: format!("Title earned: {title}"),
                style: Some(OutputStyle::Loot),
            },
        );
    }

    send(&username, ServerMessage::Bell);
}

pub fn format_achievements(player: &PlayerSession) -> String {
    if player.data.achievements.is_empty() {
        return "No achievements yet. Explore, fight, and conquer!".into();
    }

    let mut lines = vec!["-- Achievements --".into()];

    for id in &player.data.achievements {
        if let Some(a) = crate::achievements::achievements().get(id.as_str()) {
            lines.push(format!("  ✓ {} — {}", a.name, a.description));
        }
    }

    if let Some(title) = &player.data.title {
        lines.push(format!("\nTitle: {title}"));
    }

    lines.join("\n")
}